Array_Vector_Exercises/Exercise1/TDinamicArray.h

```diff
@@ -0,0 +1,242 @@
+#pragma once
+
+template<typename T>
+class TDynamicArray
+{
+private:
+	int SizeArray;
+	int CapacityArray;
+	T* DynamicArray;
+
+public:
+	TDynamicArray()
+	{
+		SizeArray = 0;
+		CapacityArray = 1;
+		DynamicArray = new T[CapacityArray];
+	};
+
+	~TDynamicArray()
+	{
+		delete[] DynamicArray;
+	};
+
+	TDynamicArray(const TDynamicArray<T>& OtherDynamicArray)
+	{
+		Append(DynamicArray);
+	}
+
+	TDynamicArray& operator= (const TDynamicArray<T>& OtherDynamicArray)
+	{
+		Clear();
+		Append(OtherDynamicArray);
+		return *this;
+	}
+
+	//----------------------------------------------------------------------
+	T& operator[] (const int Index)
+	{
+		return DynamicArray[Index];
+	}
+
+	const T& operator[] (const int Index) const
+	{
+		return DynamicArray[Index];
+	}
+
+	//-----------------------------------------------------------------------
+	T& Front()
+	{
+		return DynamicArray[0];
+	}
+
+	const T& Front() const
+	{
+		return DynamicArray[0];
+	}
+
+	//------------------------------------------------------------------------
+	T& Back()
+	{
+		return DynamicArray[SizeArray - 1];
+	}
+
+	const T& Back() const
+	{
+		return DynamicArray[SizeArray - 1];
+	}
+
+	//-----------------------------------------------------------------------
+	T* GetData()
+	{
+		return DynamicArray;
+	}
+
+	const T* GetData() const
+	{
+		return DynamicArray;
+	}
+
+	//------------------------------------------------------------------------
+	bool IsEmpty() const
+	{
+		return SizeArray == 0;
+	};
+
+	//-------------------------------------------------------------------------
+	int GetSize() const
+	{
+		return SizeArray;
+	}
+
+	//-------------------------------------------------------------------------
+	int GetCapacity() const
+	{
+		return CapacityArray;
+	}
+
+	//-------------------------------------------------------------------------
+	void Reserve(const int NewCapacity)
+	{
+		if (NewCapacity > CapacityArray)
+		{
+			T* ContainerArray = new T[NewCapacity];
+
+			for (int i = 0; i < SizeArray; i++)
+			{
+				ContainerArray[i] = DynamicArray[i];
+			}
+			delete[] DynamicArray;
+
+			DynamicArray = ContainerArray;
+			CapacityArray = NewCapacity;
+		}
+	}
+
+	//--------------------------------------------------------------------------
+	void ShrinkToFit()
+	{
+		if (CapacityArray != SizeArray)
+		{
+			T* ContainerArray = new T[SizeArray];
+
+			for (int i = 0; i < SizeArray; i++)
+			{
+				ContainerArray[i] = DynamicArray[i];
+			}
+			delete[] DynamicArray;
+
+			DynamicArray = ContainerArray;
+		}
+	}
+
+	//--------------------------------------------------------------------------
+	void Clear()
+	{
+		SizeArray = 0;
+	}
+
+	//--------------------------------------------------------------------------
+	void PushBack(const T& InType)
+	{
+		if (SizeArray == CapacityArray)
+		{
+			Reserve(CapacityArray * 2);
+			DynamicArray[SizeArray] = InType;
```

Comment:
There's duplicate code here. Why write this twice?
You can do the reserve only if the condition is met, but have the rest of the code that always needs to execute outside of the if.
```
if (SizeArray == CapacityArray)
{
    Reserve(CapacityArray * 2);
}

DynamicArray[SizeArray] = InType;
SizeArray++
```